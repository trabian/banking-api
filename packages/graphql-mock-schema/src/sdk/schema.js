import { schema } from "normalizr";

const categorySchema = new schema.Entity("categories");

export const transactionSchema = new schema.Entity(
  "transactions",
  {
    category: categorySchema
  },
  {
    processStrategy: (value, parent) => ({
      accountId: parent.id,
      ...value
    })
  }
);

export const accountSchema = new schema.Entity(
  "accounts",
  {
    transactions: [transactionSchema]
  },
  {
    processStrategy: (value, parent) => ({
      userId: parent.id,
      ...value
    })
  }
);

export const userSchema = new schema.Entity("users", {
  accounts: [accountSchema]
});
