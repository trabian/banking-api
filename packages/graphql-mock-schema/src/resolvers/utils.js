import R from "ramda";

export const convertToEnum = (root, _params, ctx, { fieldName }) =>
  R.pipe(R.prop(fieldName), R.toUpper)(root);
