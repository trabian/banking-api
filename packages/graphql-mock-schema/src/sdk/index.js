export { default as LowSDK } from "./lowdb";
