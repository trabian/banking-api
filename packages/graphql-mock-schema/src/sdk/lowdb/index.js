import R from "ramda";
import low from "lowdb/lib/fp";
import MemoryAdapter from "lowdb/adapters/Memory";

import * as accounts from "./accounts";
import * as contacts from "./contacts";
import * as transactions from "./transactions";
import * as users from "./users";

export const createSdk = async ({
  defaultValue = {
    accounts: [],
    categories: [],
    contacts: [],
    transactions: [],
    users: [],
  },
  adapter,
} = {}) => {
  const db = await Promise.resolve(
    low(
      adapter ||
        new MemoryAdapter(undefined, {
          defaultValue,
        })
    )
  );

  const sdkContext = {
    db,
  };

  return R.pipe(R.mergeAll, R.map(R.applyTo(sdkContext)))([
    accounts,
    contacts,
    transactions,
    users,
  ]);
};
