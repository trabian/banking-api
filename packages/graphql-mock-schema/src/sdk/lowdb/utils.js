import R from "ramda";

// Creates a curried finder based on the table name
export const createFinder = (tableName, filterFn = R.identity) => {
  return ({ db }) => async (...args) => db(tableName)(filterFn(...args));
};

export const createUpdater = (tableName, updateFn = R.identity) => {
  return ({ db }) => async (id, ...args) => {
    const matcher = R.propEq("id", id);

    const updateMatchingRecord = R.map(R.when(matcher, updateFn(...args)));

    const _db = db(tableName);

    await _db.write(updateMatchingRecord);

    return _db(R.find(matcher));
  };
};
