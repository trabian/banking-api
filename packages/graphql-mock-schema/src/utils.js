export const gql = str => str[0];
