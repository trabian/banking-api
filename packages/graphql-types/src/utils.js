// gql is a nice hint to editors and prettier, but we don't really want to parse the graphql schema.
export const gql = str => str[0];
