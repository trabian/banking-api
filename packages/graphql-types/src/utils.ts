// gql is a nice hint to editors and prettier, but we don't really want to parse the graphql schema.
export const gql = (literals: any): string => literals.raw[0];
