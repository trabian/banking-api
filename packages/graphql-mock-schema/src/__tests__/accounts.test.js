import { typeDefs, resolvers } from "../index";
import { graphql } from "graphql";
import { makeExecutableSchema } from "graphql-tools";
import R from "ramda";

import { gql } from "../utils.js";

const schema = makeExecutableSchema({ typeDefs, resolvers });

import { createLowSdk as createSdk } from "../sdk";

describe("The accounts schema using the LowSDK", () => {
  let execute;

  beforeEach(async () => {
    const sdk = await createSdk({
      defaultValue: {
        users: [
          {
            id: "1"
          }
        ],
        accounts: [
          {
            id: "1234",
            actualBalance: 2345.67,
            userId: "1"
          },
          {
            id: "2345",
            userId: "1",
            actualBalance: 2345.67
          },
          {
            id: "not-mine",
            actualBalance: 100,
            userId: "2"
          }
        ],
        transactions: [
          {
            accountId: "1234",
            id: "191919"
          }
        ]
      }
    });

    await sdk.db;

    execute = query => graphql(schema, query, {}, { sdk, userId: "1" });
  });

  it("should allow access to an account based on ID", async () => {
    const result = await execute(
      gql`
        {
          account(id: "1234") {
            id
            actualBalance
          }
        }
      `
    );

    expect(R.path(["data", "account", "id"], result)).toEqual("1234");

    expect(result).toMatchSnapshot();
  });

  it("should prevent access to other users' accounts based on ID", async () => {
    const result = await execute(
      gql`
        {
          account(id: "not-mine") {
            id
          }
        }
      `
    );

    expect(R.path(["data", "account", "id"], result)).toBeUndefined();
    expect(result.errors).toBeTruthy();

    expect(result).toMatchSnapshot();
  });

  it("should show all of a user's accounts", async () => {
    const result = await execute(
      gql`
        {
          me {
            accounts {
              id
            }
          }
        }
      `
    );

    expect(
      R.pipe(R.pathOr([], ["data", "me", "accounts"]), R.pluck("id"))(result)
    ).toEqual(["1234", "2345"]);

    expect(result).toMatchSnapshot();
  });

  it("should fetch transactions", async () => {
    const result = await execute(
      gql`
        {
          account(id: "1234") {
            id
            transactions {
              id
            }
          }
        }
      `
    );

    const transactions = R.path(["data", "account", "transactions"], result);

    expect(transactions.length).toBeTruthy();
  });
});
