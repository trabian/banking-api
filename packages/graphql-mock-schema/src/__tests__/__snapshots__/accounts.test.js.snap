// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The accounts schema using the LowSDK should allow access to an account based on ID 1`] = `
Object {
  "data": Object {
    "account": Object {
      "actualBalance": 2345.67,
      "id": "1234",
    },
  },
}
`;

exports[`The accounts schema using the LowSDK should prevent access to other users' accounts based on ID 1`] = `
Object {
  "data": Object {
    "account": null,
  },
  "errors": Array [
    [GraphQLError: This is not your account.],
  ],
}
`;

exports[`The accounts schema using the LowSDK should show all of a user's accounts 1`] = `
Object {
  "data": Object {
    "me": Object {
      "accounts": Array [
        Object {
          "id": "1234",
        },
        Object {
          "id": "2345",
        },
      ],
    },
  },
}
`;
